File: Java/PrimeNumber/src/Benchmark.java
public class Benchmark {

	public Benchmark(int nth) {
		long[] primes = new long[nth];
		long nthPrime = nthPrime(nth, primes);
		System.out.println(nth+"th Prime Number: "+nthPrime);
	}

	private long nthPrime(int nth, long[] primes) {
		if(primes[0]==2 && primes[1]==3){
			for(int i=2;i<primes.length;i++)
				primes[i] = nextPrime(i-1, primes);
			return primes[nth-1];
		}else{
			primes[0] = 2L; primes[1] = 3L;
			return nthPrime(nth, primes);
		}
	}

	private long nextPrime(int idx, long[] primes) {
		long prev = primes[idx];
		while(!isPrime(prev+=2, primes));
		return prev;
	}

	private boolean isPrime(long p, long[] primes) {
		long sqrt = (long)Math.sqrt(p);
		for(int i=0;primes[i]<=sqrt;i++)
			if(p%primes[i]==0) return false;
		return true;
	}

	public static void main(String[] args) {
		new Benchmark(10000000);
	}
}

File: C/PrimeNumber/src/Benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int isPrime(unsigned long p, unsigned long *primes){
	long sqrtPrime= (long)sqrt(p);
	int i;
	for(i=0;primes[i]<=sqrtPrime;i++)
		if(p%primes[i]==0) return 0;
	return 1;
}

unsigned long nextPrime(unsigned long prev, unsigned long *primes){
	while(!isPrime(prev+=2, primes));
	return prev;
}

unsigned long nthPrime(int nth, unsigned long *primes, int curIdx){
	if(primes[0] == 2 && primes[1] == 3){
		if(nth>=curIdx){
			primes[curIdx+1] = nextPrime(primes[curIdx], primes);
			return nthPrime(nth, primes, curIdx+1);
		}else return primes[nth-1];
	}
	else{
		primes[0] = 2; primes[1] = 3;
		if(nth<=2) return primes[nth-1];
		return nthPrime(nth, primes, 1);
	}
}

main(){
	int nth = 10000000;
	unsigned long *primes = (unsigned long *)malloc(nth * sizeof(unsigned long));
	unsigned long nthPrimeNumber = nthPrime(nth, primes, -1);
	printf("%dth Prime Number: %ld\n", nth, nthPrimeNumber);
	free(primes);
}
